File: tests/Tests.CSharp/DynamicValueTests.cs
using System.Collections.Immutable;
using System.Linq;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Tests.CSharp;

// DynamicValue — the C# oracle (#3 of TS/F#/C#/Rust). Mirrors the F# canonical-shape tests
// (tests/Tests.FSharp/DynamicValue.Tests.fs) so the C# impl conforms to the same shape, lazy-bind
// accessors, structural equality (incl. native Bytes), and PropertyPath semantics.
public class DynamicValueTests
{
    private static DynamicValue.Bytes Bytes(params byte[] xs) => new(ImmutableArray.Create(xs));

    private static DynamicValue.Array Arr(params DynamicValue[] xs) => new(ImmutableArray.Create(xs));

    private static DynamicValue.Object Obj(params (string Key, DynamicValue Value)[] kvs) =>
        new(kvs.Select(kv => new KeyValuePair<string, DynamicValue>(kv.Key, kv.Value)).ToImmutableArray());

    // { "a": { "b": [10, 20, 30], "n": null }, "flag": true }
    private static readonly DynamicValue Sample =
        Obj(
            ("a", Obj(
                ("b", Arr(new DynamicValue.Int(10L), new DynamicValue.Int(20L), new DynamicValue.Int(30L))),
                ("n", new DynamicValue.Null()))),
            ("flag", new DynamicValue.Bool(true)));

    [Fact]
    public void TypeOfReportsEachShape()
    {
        Assert.Equal(DynamicValueType.Null, new DynamicValue.Null().Type);
        Assert.Equal(DynamicValueType.Bool, new DynamicValue.Bool(true).Type);
        Assert.Equal(DynamicValueType.Int, new DynamicValue.Int(1L).Type);
        Assert.Equal(DynamicValueType.Float, new DynamicValue.Float(1.0).Type);
        Assert.Equal(DynamicValueType.String, new DynamicValue.String("x").Type);
        Assert.Equal(DynamicValueType.Bytes, Bytes(1).Type);
        Assert.Equal(DynamicValueType.Array, Arr().Type);
        Assert.Equal(DynamicValueType.Object, Obj().Type);
    }

    [Fact]
    public void IsNullOnlyForNull()
    {
        Assert.True(new DynamicValue.Null().IsNull);
        Assert.False(new DynamicValue.Bool(false).IsNull);
    }

    [Fact]
    public void TryAccessorsBindMatchingShapeAndDeclineTheRest()
    {
        Assert.True(new DynamicValue.Bool(true).TryBool()!.Value);
        Assert.Null(new DynamicValue.Int(1L).TryBool());

        Assert.Equal(7L, new DynamicValue.Int(7L).TryInt()!.Value);
        Assert.Null(new DynamicValue.Float(7.0).TryInt());

        Assert.Equal(2.5, new DynamicValue.Float(2.5).TryFloat()!.Value);
        Assert.Null(new DynamicValue.Int(2L).TryFloat()); // strict: no widening

        Assert.Equal("hi", new DynamicValue.String("hi").TryString());
        Assert.Null(new DynamicValue.Null().TryString());

        Assert.True(Bytes(9).TryBytes().HasValue);
        Assert.False(new DynamicValue.String("9").TryBytes().HasValue);

        Assert.True(Arr().TryArray().HasValue);
        Assert.False(Obj().TryArray().HasValue);

        Assert.True(Obj().TryObject().HasValue);
        Assert.False(Arr().TryObject().HasValue);
    }

    [Fact]
    public void TryFieldAndTryItem()
    {
        Assert.Equal(new DynamicValue.Bool(true), Sample.TryField("flag"));
        Assert.Null(Sample.TryField("missing"));
        Assert.Null(new DynamicValue.Int(1L).TryField("flag"));

        var arr = Arr(new DynamicValue.Int(10L), new DynamicValue.Int(20L));
        Assert.Equal(new DynamicValue.Int(20L), arr.TryItem(1));
        Assert.Null(arr.TryItem(5)); // out of range
        Assert.Null(arr.TryItem(-1)); // negative
        Assert.Null(Obj().TryItem(0)); // not an array
    }

    [Fact]
    public void StructuralEqualityRecurses()
    {
        var a = Obj(("k", Arr(new DynamicValue.Int(1L), new DynamicValue.String("x"))));
        var b = Obj(("k", Arr(new DynamicValue.Int(1L), new DynamicValue.String("x"))));
        Assert.Equal(a, b);
        Assert.Equal(a.GetHashCode(), b.GetHashCode());
    }

    [Fact]
    public void BytesCompareByContentNotReference()
    {
        var a = new DynamicValue.Bytes(ImmutableArray.Create<byte>(1, 2, 3));
        var b = new DynamicValue.Bytes(ImmutableArray.Create<byte>(1, 2, 3));
        Assert.Equal(a, b);
        Assert.Equal(a.GetHashCode(), b.GetHashCode());

        var c = new DynamicValue.Bytes(ImmutableArray.Create<byte>(1, 2, 4));
        Assert.NotEqual(a, c);
    }

    [Fact]
    public void ObjectEqualityIsOrderSensitive()
    {
        var ab = Obj(("a", new DynamicValue.Int(1L)), ("b", new DynamicValue.Int(2L)));
        var ba = Obj(("b", new DynamicValue.Int(2L)), ("a", new DynamicValue.Int(1L)));
        Assert.NotEqual(ab, ba);
    }

    [Fact]
    public void DifferentShapesAreNotEqual()
    {
        Assert.NotEqual(new DynamicValue.Int(1L), (DynamicValue)new DynamicValue.Float(1.0));
        Assert.NotEqual(new DynamicValue.String("1"), (DynamicValue)new DynamicValue.Int(1L));
        Assert.NotEqual(new DynamicValue.Null(), (DynamicValue)new DynamicValue.Bool(false));
    }

    [Fact]
    public void GetNavigatesDottedAndIndexedPaths()
    {
        Assert.Equal(new DynamicValue.Int(20L), Sample.Get("a.b[1]"));
        Assert.Equal(new DynamicValue.Bool(true), Sample.Get("flag"));
        Assert.Equal(new DynamicValue.Null(), Sample.Get("a.n"));
    }

    [Fact]
    public void GetReturnsTheValueItselfForAnEmptyPath()
    {
        Assert.Equal(Sample, Sample.Get(""));
    }

    [Fact]
    public void GetDeclinesMissOutOfRangeAndTypeMismatch()
    {
        Assert.Null(Sample.Get("a.missing")); // key absent
        Assert.Null(Sample.Get("a.b[9]")); // index out of range
        Assert.Null(Sample.Get("flag.x")); // descend into a non-object
    }

    [Fact]
    public void GetSupportsABareLeadingIndex()
    {
        var arr = Arr(new DynamicValue.String("zero"), new DynamicValue.String("one"));
        Assert.Equal(new DynamicValue.String("one"), arr.Get("[1]"));
    }

    [Fact]
    public void GetDeclinesMalformedPaths()
    {
        Assert.Null(Sample.Get("a.b[")); // unterminated bracket
        Assert.Null(Sample.Get("a.b[x]")); // non-digit index
        Assert.Null(Sample.Get("a]b")); // stray close bracket
        Assert.Null(Sample.Get("a.b[99999999999999999999]")); // index overflows Int32 -> null, not an exception
    }
}

File: tests/Tests.FSharp/DynamicValue.Tests.fs
module Zeta.Tests.DynamicValueTests

open System.Collections.Immutable
open global.Xunit
open Zeta.Core

// DynamicValue — the F# canonical reference shape for the format-agnostic
// dynamic / self-describing-payload primitive. These tests fix the shape's
// behaviour (tag, lazy-bind accessors, structural equality incl. native Bytes,
// PropertyPath navigation) before the C#/Rust/TS oracles conform to it.

let private bytes (xs: byte list) : DynamicValue =
    DynamicValue.Bytes(ImmutableArray.CreateRange xs)

// Sample tree: { "a": { "b": [10, 20, 30], "n": null }, "flag": true }
let private sample : DynamicValue =
    DynamicValue.Object
        [ "a",
          DynamicValue.Object
              [ "b", DynamicValue.Array [ DynamicValue.Int 10L; DynamicValue.Int 20L; DynamicValue.Int 30L ]
                "n", DynamicValue.Null ]
          "flag", DynamicValue.Bool true ]

// -- typeOf (the QueryInterface tag) --

[<Fact>]
let ``typeOf reports each shape`` () =
    Assert.Equal(DynamicValueType.Null, DynamicValue.typeOf DynamicValue.Null)
    Assert.Equal(DynamicValueType.Bool, DynamicValue.typeOf (DynamicValue.Bool true))
    Assert.Equal(DynamicValueType.Int, DynamicValue.typeOf (DynamicValue.Int 1L))
    Assert.Equal(DynamicValueType.Float, DynamicValue.typeOf (DynamicValue.Float 1.0))
    Assert.Equal(DynamicValueType.String, DynamicValue.typeOf (DynamicValue.String "x"))
    Assert.Equal(DynamicValueType.Bytes, DynamicValue.typeOf (bytes [ 1uy ]))
    Assert.Equal(DynamicValueType.Array, DynamicValue.typeOf (DynamicValue.Array []))
    Assert.Equal(DynamicValueType.Object, DynamicValue.typeOf (DynamicValue.Object []))

[<Fact>]
let ``isNull only for Null`` () =
    Assert.True(DynamicValue.isNull DynamicValue.Null)
    Assert.False(DynamicValue.isNull (DynamicValue.Bool false))

// -- lazy bind accessors (strict; no widening) --

[<Fact>]
let ``try accessors bind matching shape and decline the rest`` () =
    Assert.Equal(Some true, DynamicValue.tryBool (DynamicValue.Bool true))
    Assert.Equal(None, DynamicValue.tryBool (DynamicValue.Int 1L))

    Assert.Equal(Some 7L, DynamicValue.tryInt (DynamicValue.Int 7L))
    Assert.Equal(None, DynamicValue.tryInt (DynamicValue.Float 7.0))

    Assert.Equal(Some 2.5, DynamicValue.tryFloat (DynamicValue.Float 2.5))
    Assert.Equal(None, DynamicValue.tryFloat (DynamicValue.Int 2L)) // strict: no widening

    Assert.Equal(Some "hi", DynamicValue.tryString (DynamicValue.String "hi"))
    Assert.Equal(None, DynamicValue.tryString DynamicValue.Null)

    Assert.True((DynamicValue.tryBytes (bytes [ 9uy ])).IsSome)
    Assert.Equal(None, DynamicValue.tryBytes (DynamicValue.String "9"))

    Assert.True((DynamicValue.tryArray (DynamicValue.Array [])).IsSome)
    Assert.Equal(None, DynamicValue.tryArray (DynamicValue.Object []))

    Assert.True((DynamicValue.tryObject (DynamicValue.Object [])).IsSome)
    Assert.Equal(None, DynamicValue.tryObject (DynamicValue.Array []))

[<Fact>]
let ``tryField and tryItem`` () =
    Assert.Equal(Some(DynamicValue.Bool true), DynamicValue.tryField "flag" sample)
    Assert.Equal(None, DynamicValue.tryField "missing" sample)
    Assert.Equal(None, DynamicValue.tryField "flag" (DynamicValue.Int 1L))

    let arr = DynamicValue.Array [ DynamicValue.Int 10L; DynamicValue.Int 20L ]
    Assert.Equal(Some(DynamicValue.Int 20L), DynamicValue.tryItem 1 arr)
    Assert.Equal(None, DynamicValue.tryItem 5 arr) // out of range
    Assert.Equal(None, DynamicValue.tryItem -1 arr) // negative
    Assert.Equal(None, DynamicValue.tryItem 0 (DynamicValue.Object [])) // not an array

// -- structural equality (hand-written; Bytes compares contents) --

[<Fact>]
let ``structural equality recurses`` () =
    let a =
        DynamicValue.Object [ "k", DynamicValue.Array [ DynamicValue.Int 1L; DynamicValue.String "x" ] ]

    let b =
        DynamicValue.Object [ "k", DynamicValue.Array [ DynamicValue.Int 1L; DynamicValue.String "x" ] ]

    Assert.Equal(a, b)
    Assert.Equal(a.GetHashCode(), b.GetHashCode())

[<Fact>]
let ``Bytes compare by content not reference`` () =
    // Two distinct ImmutableArray instances with identical bytes must be equal.
    let a = DynamicValue.Bytes(ImmutableArray.Create<byte>(1uy, 2uy, 3uy))
    let b = DynamicValue.Bytes(ImmutableArray.Create<byte>(1uy, 2uy, 3uy))
    Assert.Equal(a, b)
    Assert.Equal(a.GetHashCode(), b.GetHashCode())

    let c = DynamicValue.Bytes(ImmutableArray.Create<byte>(1uy, 2uy, 4uy))
    Assert.NotEqual(a, c)

[<Fact>]
let ``Object equality is order-sensitive`` () =
    let ab =
        DynamicValue.Object [ "a", DynamicValue.Int 1L; "b", DynamicValue.Int 2L ]

    let ba =
        DynamicValue.Object [ "b", DynamicValue.Int 2L; "a", DynamicValue.Int 1L ]

    // The value tree preserves insertion order; canonical encoders sort on the wire.
    Assert.NotEqual(ab, ba)

[<Fact>]
let ``different shapes are not equal`` () =
    Assert.NotEqual(DynamicValue.Int 1L, DynamicValue.Float 1.0)
    Assert.NotEqual(DynamicValue.String "1", DynamicValue.Int 1L)
    Assert.NotEqual(DynamicValue.Null, DynamicValue.Bool false)

// -- PropertyPath navigation --

[<Fact>]
let ``get navigates dotted and indexed paths`` () =
    Assert.Equal(Some(DynamicValue.Int 20L), DynamicValue.get "a.b[1]" sample)
    Assert.Equal(Some(DynamicValue.Bool true), DynamicValue.get "flag" sample)
    Assert.Equal(Some DynamicValue.Null, DynamicValue.get "a.n" sample)

[<Fact>]
let ``get returns the value itself for an empty path`` () =
    Assert.Equal(Some sample, DynamicValue.get "" sample)

[<Fact>]
let ``get declines miss, out-of-range, and type-mismatch`` () =
    Assert.Equal(None, DynamicValue.get "a.missing" sample) // key absent
    Assert.Equal(None, DynamicValue.get "a.b[9]" sample) // index out of range
    Assert.Equal(None, DynamicValue.get "flag.x" sample) // descend into a non-object

[<Fact>]
let ``get supports a bare leading index`` () =
    let arr = DynamicValue.Array [ DynamicValue.String "zero"; DynamicValue.String "one" ]
    Assert.Equal(Some(DynamicValue.String "one"), DynamicValue.get "[1]" arr)

[<Fact>]
let ``get declines malformed paths`` () =
    Assert.Equal(None, DynamicValue.get "a.b[" sample) // unterminated bracket
    Assert.Equal(None, DynamicValue.get "a.b[x]" sample) // non-digit index
    Assert.Equal(None, DynamicValue.get "a]b" sample) // stray close bracket
    Assert.Equal(None, DynamicValue.get "a.b[99999999999999999999]" sample) // index overflows Int32 -> None, not an exception

[<Fact>]
let ``get declines empty path segments`` () =
    Assert.Equal(None, DynamicValue.get ".flag" sample) // leading dot
    Assert.Equal(None, DynamicValue.get "a..b" sample) // doubled dot
    Assert.Equal(None, DynamicValue.get "a." sample) // trailing dot
    Assert.Equal(None, DynamicValue.get "." sample) // lone dot

[<Fact>]
let ``default ImmutableArray bytes behave as empty`` () =
    // A default (uninitialized) ImmutableArray must not poison equality/hashing/accessors —
    // it normalizes to the empty payload rather than throwing on enumeration.
    let d = DynamicValue.Bytes Unchecked.defaultof<ImmutableArray<byte>>
    let e = DynamicValue.Bytes ImmutableArray<byte>.Empty
    Assert.Equal(d, e)
    Assert.Equal(d.GetHashCode(), e.GetHashCode())
    Assert.True((DynamicValue.tryBytes d).Value.IsEmpty)
